File: diffusivity/kmodel.py
#!/usr/bin/env python
## Time-stamp: "2015-12-03 17:03:20 marine"

import numpy as np
from scipy.integrate import quad
import matplotlib.pyplot as plt


IRON = {'name' : 'Fe', 'f1' : 5.26e-9, 'f2' : 1.24, 'f3' : -3.21}

def rho(f,f1,f2,f3) :
    """ equation 9 Gomi 2013 """
    return f1*(f2 - f)**f3
    


def resistivity(f, T, El, theta0=417., gamma=1.52, Ir=IRON) :
    """ Ideal resistivity from the Bloch GRuneisen formula as in Gomi 2013 (equation (1))
    
    f: V/V0
    T : absolute temperature
    El: element (dictionnary with z valence charge and x atomic fraction)
    theta0
    gamma
    
    """

    def integrand_eq1(z):
        """ integrande for function resistivity """
        return z**5/(np.exp(z)-1)/(1-np.exp(-z))

    theta = theta0*np.exp(-gamma *np.log(f))  #Debye temperature
    tfac = (T/theta)**5 *quad(integrand_eq1, 0, theta/T)[0]
    
    imp = El['z']**2*El['x']

    idSi = rho(f, El['f1'], El['f2'], El['f3'])
    idFe = rho(f, Ir['f1'], Ir['f2'], Ir['f3']) * tfac

    return idFe + idSi*imp  #ideal resistivity


def k(id, T, f, L=2.44e-8):
    """ real resistivity, adding ideal resistivity (id) and saturation resistivity (sat)
    saturation resistivity at 1bar is 1.68e-6 \Ohm m
    """
    sat = 1.68e-6 * f**(1./3.)
    return L*T*(1./id+ 1./sat)



if __name__ == '__main__':

    import eos

    ## Light elements :
    # name, z: valence charge, x: atomic fraction
    # f1,f2,f3 are fitting parameters
    IRON = {'name' : 'Fe', 'f1' : 5.26e-9, 'f2' : 1.24, 'f3' : -3.21}
    SILICIUM = {'name' : 'Si', 'z' : 1, 'x' : 22.5, 'f1' : 3.77e-8, 'f2' : 1.48, 'f3' : -3.10}
    CARBON = {'name' : 'C', 'z' : 1, 'x' : 30}
    OXYGEN = {'name' : 'O', 'z' : 0.5, 'x' : 23.2}
    SULFUR = {'name' : 'S', 'z' : 0.5, 'x' : 19.4}
    
    MURNAGHAN = {'T0': 1812., 'rho0': 7010., 'K': 130.*1e9, 'KPrim':4.,
                        'alphaPrim': 130.*np.log(2.)/(360.-135.) ,
                        'alpha': 1.e-5/np.exp(-0.400485 *135./130.),   #1.e-5*np.exp(130.*np.log(2.)/(360.-135.)*135./130),#   1.e-5/np.exp( np.log(2)/(360.-135.)*135.),
                        'Cp': 800., 'cond': 140.}
    P0 = 135e9
    T0 = 3750.

    N = 100 #number of points in the profile
    P = np.linspace(P0,330.e9,N)
    T = []

    ## Adiabatic profile
    for p in P:
        #print Tadiabatic(p, T0, P0, MURNAGHAN)
        T = np.append(T, eos.Tadiabatic(p, T0, P0, MURNAGHAN))

        
    fig1 = plt.figure()
    ax1 = fig1.add_subplot(111)
    ax1.plot(P/1.e9,T)
    ax1.set_xlabel("Pressure (GPa)")
    ax1.set_xlabel("Temperature (K)")
    ax1.set_title("Adiabatic profile")

    ## Associated k profile
    conductivity = []
    for i in range(0,N):        
        f = eos.fval(P[i],T[i], MURNAGHAN)
        res = resistivity(f, T[i], El=SILICIUM, theta0=417, gamma=1.52, Ir=IRON)
        conductivity = np.append(conductivity, k(res, T[i], f, L=2.44e-8))

    fig2 = plt.figure()
    ax2 = fig2.add_subplot(111)
    ax2.plot(P/1.e9,conductivity)
    ax2.set_xlabel("Pressure (GPa)")
    ax2.set_xlabel("Conductivity")
    ax2.set_title("Conductivity profile along the adiabatic profile")

    
    plt.show()

File: diffusivity/kmodel.R
## Gomi et al. (2013) model for conductivity in the core.

el.si<- list( list(z=1,  x=22.5,name='Si') )      ## Si as light element
el.c <- list( list(z=1,  x=30,  name='C') )       ## C as light element
el.o <- list( list(z=0.5,x=23.2,name='O') )       ## O as light element
el.s <- list( list(z=0.5,x=19.4,name='S') )       ## S as light element

el <- el.si          ## Change this assignment for different light element

res <- function(f, T, zx=list(), theta0=417, gamma=1.52){
   ## f - V/V0
   ## T - absolute temperature
   ## zx - vector of impurities, each a list:
   ##   z= valence charge, x= concentration (atomic fraction)

   theta <- theta0*exp(-gamma*log(f))
   rho <- function(f, f1, f2, f3){f1*(f2 - f)^f3}

   tfac <- integrate(
      function(z){z^5/((exp(z)-1)*(1-exp(-z)))},
      0, theta/T
   )$value * (T/theta)^5

   id.fe <- rho(f, 5.26e-9, 1.24, -3.21)*tfac
   id.si <- rho(f, 3.77e-8, 1.48, -3.10)
   imp <- 0; for(el in zx){
      imp <- imp + el$z^2*el$x
   }
   id.fe + id.si*imp
}

k <- function(id,T,f, L=2.44e-8){
   ## id - ideal resistivity
   ## T - absolute temperature
   ## f - V/V0

   sat <- 1.68e-6 * f^(1/3)

   (L*T)*(1/id + 1/sat)
}

## EOS functions; generic, could be replaced by more specific ones.

eos <- function(p,t,meos){
   ## Murnaghan equation of state
   ## p - pressure in Pa
   ## t - absolute temperature
   ## meos - material properties list for EOS
   rho.p <- meos$rho0*(meos$k.p * p/meos$k + 1)^(1/meos$k.p)
   al.pt <- exp(-(meos$al*exp(-meos$al.p/meos$k*p))*(t - meos$t0))
   # cat('p,t,rho.p,al.pt',p,t,rho.p,al.pt,'\n')
   rho.p*al.pt
}

eos.al <- function(p,meos){
   meos$al*exp(-meos$al.p/meos$k*p)
}

eos.tad <- function(p,t0,p0,meos){
   ## Return adiabatic lapse from t0, p0 to p
   ## t0 - absolute T at P=p0
   ## p0 - P0 pressure (Pa)
   ## p - pressure (Pa)

   t0*exp(integrate(function(p){
      eos.al(p,meos)/eos(p,t0,meos)
   }, p0, p)$value / meos$Cp)
}

# Calculate alpha such that value at 135 GPa is 1e-5, 360 GPa is 0.5e-5
rho0.met <- 7010; k.met <- 130         # Literature values
al.p <- k.met*log(2)/(360-135); al <- 1e-5/exp(-al.p*135/k.met)
meos.met <- list(
   t0=1812, rho0=rho0.met, k=k.met*1e9, k.p=4, al=al, al.p=al.p,
   Cp=800, cond=140
)

fval <- function(P, T, meos=meos.met){
   ## Return f = V/V0 for material
   ## P - pressure in GPa (note)
   ## T - absolute temperature
   eos(0,300,meos)/eos(P*1e9,T,meos)
} 

## Test to calculate variation of k along an adiabat rooted at the CMB with
##   temperature T=3750 K and P=135 GPa to ICB P=330 GPa

p <- seq(135,330,5)

t <- sapply(p, function(p)eos.tad(p*1e9, 3750, 135e9, meos.met)) ## adiabat

print(meos.met)

plot(p, t, type='l', xlab='P (GPa)', ylab='T (K)')  ## Plot T along adiabat

kvals <- sapply(1:length(p),function(i){            ## Calculate k along adiabat
   f <- fval(p[i], t[i])
   rho <- res(f, t[i], el)
   k(rho, t[i], f)
})

dev.new(); spar <- par('mar'); par(mar=0.1+c(5,5,4,2)) ## Extra margin room

plot(p, kvals, type='l',                            ## Plot k along adiabat
   xlab='P (GPa)',
   ylab=expression(k~(W~m^{-1}~K^{-1}))
)
text(p[1]+0.1*diff(range(p)), mean(kvals),
   sprintf("   %s (%.1f at.%%)",el[[1]]$name,el[[1]]$x)
)

par(spar)
